### PR TITLE
#18 - Systems with outputs

### LINKED ISSUE #18: LTI systems type
In control books, (continuous) LTI systems are those of the form: `x' = Ax + Bu; y = Cx + Du`.

Proposal: define a `LinearTimeInvariantContinuousSystem` with fields `A, B, C, D` and another one `ConstrainedLinearTimeInvariantContinuousSystem` with fields `A, B, C, D, X, U`.

### PR DESCRIPTION
**Pull request by mforets**
Closes #18.

### CODE REVIEW

**Comment by mforets** [comment]
i've hesitated about `...maps` because one can argue that `...DiscreteSystem` can be used for that purpose. perhaps we can make the `MT` in the output be optional, eg.

```julia
struct SystemWithOutput{ST<:AbstractSystem, MT}   # no restriction on MT
    system::ST
    outputmap::MT
end
```

File: src/maps.jl [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+- `U` -- input constraints
+"""
+struct LinearControlMap{MT<:AbstractMatrix, UT} <: AbstractMap
+    A::MT
+    B::MT
+    U::UT
+end
+
+"""
+    AffineMap
```
Comment (schillic) [inline src/maps.jl]:
```suggestion
    AffineControlMap
```

**Comment by schillic** [comment]
I do not understand what you want to say, sorry.

File: src/maps.jl [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+An affine map
```
Comment (blegat) [inline src/maps.jl]: affine control

File: src/maps.jl [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+"""
+struct AffineControlMap{MT<:AbstractMatrix, VT<:AbstractVector, UT} <: AbstractMap
+    A::MT
+    B::MT
+    c::VT
+    U::UT
+end
+
+"""
+    SystemWithOutput
```
Comment (blegat) [inline src/maps.jl]: Maybe add parameters here and add in the `Fields` section "of type `ST`"

File: src/maps.jl [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+"""
+struct AffineMap{MT<:AbstractMatrix, VT<:AbstractVector} <: AbstractMap
+    A::MT
+    b::VT
+end
+
+"""
+    LinearControlMap
+
+An affine map
```
Comment (blegat) [inline src/maps.jl]: affine -> linear control

**Comment by schillic** [comment]
>```julia
>struct LinearMap{T, MT<:AbstractMatrix} <: AbstractMap
>    A::MT
>end
>```

You need to use `T` here :wink:

**Comment by blegat** [comment]
Yes it should be `AbstractMatrix{T}`

**Comment by mforets** [comment]
i've added 2 more types `ConstrainedLinearControlMap` and `ConstrainedAffineControlMap` making `LinearControlMap` and `AffineControlMap` to be unconstrained (as it is the case for systems).

then i've added unit tests and docs.

new methods `outputdim` and `outputmap` have been defined for a `SystemWithOutput`. 

an LTI system can be constructed as a `LTISystem(A, B, C, D)` or a `LTISystem(A, B, C, D, X, U)` if one wants to specify both state and input constraints.

ready from my side.

File: src/MathematicalSystems.jl [hunk 5]
```diff
@@ -14,7 +14,8 @@ export AbstractSystem,
        statedim,
        stateset,
        inputdim,
-       inputset
+       inputset,
+       outputdim
```
[the remark this answers is not in the document]
Comment (mforets) [reply src/MathematicalSystems.jl]: ok

File: src/outputs.jl [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+- `D` -- matrix
+
+### Output
+
+A system with output such that the system is a linear control
+continuous system and the output map is a linear control map.
+"""
+function LinearTimeInvariantSystem(A, B, C, D)
+    system = LinearControlContinuousSystem(A, B)
+    outputmap = LinearControlMap(A, B)
```
Comment (blegat) [inline src/outputs.jl]: It should be `C`, `D` here

Comment (mforets) [reply src/outputs.jl]: oops

Comment (blegat) [reply src/outputs.jl]: We should add a test to catch that

File: test/outputs.jl [hunk 7]
```diff
@@ -2,21 +2,25 @@
...
     s = LinearTimeInvariantSystem(A, B, C, D)
     @test statedim(s) == 2
     @test inputdim(s) == 1
     @test outputmap(s) isa LinearControlMap
 
     # check alias
     s = LTISystem(A, B, C, D)
+
+    # check the LTI system's outputmap fields 
+    @test outputmap(s).A == C
```
Comment (schillic) [inline test/outputs.jl]: You could add the analogous tests for `s.system`.

**Pull request merged**